### PR TITLE
bug: Fix MultilineWhitespaceBeforeSemicolonsFixer

### LINKED ISSUE #4896: Unable to multiline_whitespace_before_semicolons with both strategies
`multiline_whitespace_before_semicolons` is a great fixer, but it's unclear how to use it with both strategies.
I want to forbid multi-line whitespace before the closing semicolon except chained calls.
I get this result if I run a fixer with the `no_multi_line` strategy, and then run it again with the `new_line_for_chained_calls` strategy, but this cannot be done in one pass.

Are there any technical limitations to do this, or have I just not figured out how to do this?

#### The PHP version you are using (`$ php -v`):
```
PHP 7.4.3 (cli) (built: Feb 20 2020 21:53:46) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.3, Copyright (c), by Zend Technologies
    with Xdebug v2.9.2, Copyright (c) 2002-2020, by Derick Rethans
```

#### PHP CS Fixer version you are using (`$ php-cs-fixer -V`):
```
PHP CS Fixer 2.16.1 Yellow Bird by Fabien Potencier and Dariusz Ruminski
```

#### The command you use to run PHP CS Fixer:
```
vendor/bin/php-cs-fixer fix -vv --using-cache=no
```

#### The configuration file you are using, if any:
```php
$finder = PhpCsFixer\Finder::create()
    ->in(__DIR__)
    ->exclude('var')
;

return PhpCsFixer\Config::create()
    ->setRules([
        'multiline_whitespace_before_semicolons' => ['strategy' => 'new_line_for_chained_calls'],
    ])
    ->setFinder($finder)
;
```

#### If applicable, please provide minimum samples of PHP code (as plain text, not screenshots):
 * before running PHP CS Fixer (no changes):
```php
$this
    ->foo()
    ->bar();
$this->baz()
;
```

 * with unexpected changes applied when running PHP CS Fixer:
with `new_line_for_chained_calls` strategy:
```php
$this
    ->foo()
    ->bar()
;
$this->baz()
;
```
with `no_multi_line` strategy:
```php
$this
    ->foo()
    ->bar();
$this->baz();
```

 * with the changes you expected instead:
```php
$this
    ->foo()
    ->bar()
;
$this->baz();
```

### PR DESCRIPTION
**Pull request by ossinkine**
Closes #4896

Now `new_line_for_chained_calls` do the same fixes as `no_multi_line` except chaining calls.

Also it fixes wrong idententions in chainig calls:
```php
// before
$foo
    ->bar()
    ;

// after
$foo
    ->bar()
;
```

It also seemed to me that `no_multi_line` fixes with comments were not working correctly. I fixed it, but if the reviewers decide that it should be returned, I will:
```php
//before
    $foo->bar() // test
;
                    
//after
    $foo->bar(); // test
```

### CODE REVIEW

File: src/Console/Application.php [hunk 1]
```diff
@@ -78,8 +78,7 @@ public function doRun(InputInterface $input, OutputInterface $output)
     {
         $stdErr = $output instanceof ConsoleOutputInterface
             ? $output->getErrorOutput()
-            : ($input->hasParameterOption('--format', true) && 'txt' !== $input->getParameterOption('--format', null, true) ? null : $output)
```
Comment (julienfalque) [inline src/Console/Application.php:81]: IMO this change should be reverted, i.e. the semicolon should remain on a newline. Same with similar changes in other files.

Comment (ossinkine) [reply src/Console/Application.php:81]: Why? This is behavior of `no_multi_line` strategy and this is not chaned calls.

Comment (julienfalque) [reply src/Console/Application.php:81]: Ah right, I was actually expressing a feature request, out of scope here :)

**Comment by ossinkine** [comment]
@julienfalque @SpacePossum @mvorisek @kubawerlos @GrahamCampbell
Hey guys, could someone review this?

**Comment by kubawerlos** [comment]
Travis CI already did: https://travis-ci.org/github/FriendsOfPHP/PHP-CS-Fixer/jobs/731294846

There is no point in review until this is solved.

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 2]
```diff
@@ -697,8 +697,7 @@ private function injectAlignmentPlaceholdersForArrow(Tokens $tokens, $startAt, $
                     if ($tokens[$i + 1]->isGivenKind([T_ARRAY, CT::T_ARRAY_SQUARE_BRACE_OPEN])) {
                         $arrayStartIndex = $tokens[$i + 1]->isGivenKind(T_ARRAY)
                             ? $tokens->getNextMeaningfulToken($i + 1)
-                            : $i + 1
-                        ;
+                            : $i + 1;
```
Comment (GrahamCampbell) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php:700]: I don't think these kind of changes are wanted?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php (hunk 2 above)
Comment (ossinkine) [reply src/Fixer/Operator/BinaryOperatorSpacesFixer.php:700]: Why? This is what the fixer do, I've just fix the behavior for chaining calls, see #4896

File: src/Fixer/Semicolon/MultilineWhitespaceBeforeSemicolonsFixer.php [hunk 3]
```diff
@@ -107,76 +108,66 @@ protected function createConfigurationDefinition()
...
-            $content = $previous->getContent();
-            if (0 === strpos($content, $lineEnding) && $tokens[$index - 2]->isComment()) {
-                $tokens->ensureWhitespaceAtIndex($previousIndex, 0, $lineEnding);
-            } else {
-                $tokens->clearAt($previousIndex);
-            }
-        }
-    }
+            if (self::STRATEGY_NEW_LINE_FOR_CHAINED_CALLS === $this->configuration['strategy']
+                && null !== ($indent = $this->findWhitespaceBeforeFirstCall($index - 1, $tokens))
```
Comment (mvorisek) [inline src/Fixer/Semicolon/MultilineWhitespaceBeforeSemicolonsFixer.php:123]: do no assign in `if`, see Travis error

Comment (ossinkine) [reply src/Fixer/Semicolon/MultilineWhitespaceBeforeSemicolonsFixer.php:123]: Yes, thank you, I'll do it shortly

**Comment by ossinkine** [comment]
@kubawerlos @mvorisek fixed

**Comment by ossinkine** [comment]
@kubawerlos @mvorisek Hey guys, tests are passed, let's merge this if everything is ok

**Comment by GrahamCampbell** [comment]
Can you regression test this on symfony/symfony to check there are no crashes or bad changes? If that is all good, then 👍 from me.

**Comment by ossinkine** [comment]
@GrahamCampbell I've ran the fixer on symfony/symfony and there was no fixes, but I guess it's so because Symfony doesn't use this fixer

**Comment by GrahamCampbell** [comment]
>  I've ran the fixer on symfony/symfony and there was no fixes, but I guess it's so because Symfony doesn't use this fixer

Don't use their config file.

**Comment by ossinkine** [comment]
@GrahamCampbell I've ran this fixer on symfony/symfony and found some issues, fixed them and added tests. Please review again.

**Comment by GrahamCampbell** [comment]
Could you add some more tests with super funky comments all over the place, so we can check in particular that a semicolon doesn't get moved onto a line with a `//` comment, breaking the code.

**Comment by ossinkine** [comment]
There are already a lot of cases with comments, so please specify what exactly cases you want to add. Like the following?
```php
$this
    ->first() // first
    ->seconds() //second
    ->third() // third
;
```

Actually the issue is not with comments position but with chaining syntax like that I covered by tests.

**Comment by mvorisek** [comment]
@ossinkine please look into `no_multi_line`, I think this option is still broken.

Steps to reproduce:
1. clone https://github.com/atk4/dsql.git
2. edit `.php_cs.dist` `'multiline_whitespace_before_semicolons' => false,` line to `'multiline_whitespace_before_semicolons' => ['strategy' => 'no_multi_line'],`
3. run fixer
4. check if https://github.com/atk4/dsql/blob/ece86bbecf33ebc44955a877cf25b4d87b0ddd9c/tests/QueryTest.php#L1581 is fixed correctly - that semicolon should be moved at the end of the previous line before comment

**Comment by ossinkine** [comment]
@mvorisek Looks like the fixer works correctly on your case:
```diff
--- Original
+++ New
@@ @@
             ->field('amount', 'debit')
-            ->field($this->q()->expr('0'), 'credit') // simply 0
-;
+            ->field($this->q()->expr('0'), 'credit'); // simply 0
@@ @@
             ->option('calc_found_rows', 'select') // for default select mode
-            ->option('ignore', 'insert') // for insert mode
-;
+            ->option('ignore', 'insert'); // for insert mode
```

**Review by GrahamCampbell — APPROVED** [review]

**Review by mvorisek — APPROVED** [review]
Tested, 👍

**Review by kubawerlos — APPROVED** [review]

**Comment by GrahamCampbell** [comment]
@SpacePossum RTM?

**Comment by ossinkine** [comment]
@keradus why we can't merge it and you fourth time move the milestone?

**Comment by robfrawley** [comment]
@keradus Is this PR still expected to be merged at some point? Was an alternate implementation merged instead? Or have the project maintainers decided against this change?

If it is still intended to be merged, it would be great if this could be included in the upcoming `3.0.0` release.

**Comment by julienfalque** [comment]
Not sure I would merge this PR yet, it improves the behavior in some cases but prevents others from being allowed, e.g. this does not seem to be allowed anymore:
```php
return
    $foo
    && $bar
;
```
or
```php
return $foo
    ? $bar
    : null
;
```
Although I think #3679 supports those cases so maybe it's fine.

**Comment by kubawerlos** [comment]
@ossinkine we need to solve the conflicts.

**Comment by kubawerlos** [comment]
Also, can you target `3.0` branch as this is a bugfix?

**Comment by ossinkine** [comment]
@kubawerlos I've solved the conflicts and changed the target branch, let's merge

**Comment by ossinkine** [comment]
@keradus you've changed the target branch to `master`, but @kubawerlos requested to change to `3.0`. What is correct branch?

**Comment by kubawerlos** [comment]
@ossinkine try to change it back 😉

**Comment by ossinkine** [comment]
@kubawerlos There is the branch `master` only, `3.0` does not exist anymore

**Comment by kubawerlos** [comment]
Exactly, can you solve the conflicts? Would be if not RTM, if not RTM as already have 3 approves.

**Comment by ossinkine** [comment]
@kubawerlos hope this was a final rebase

**Review by kubawerlos — APPROVED** [review]
Looking very good 👍🏼

**Comment by ossinkine** [comment]
There are conflicts again. Could we merge it right after my rebasing before new conflicts appear?

**Comment by ossinkine** [comment]
@julienfalque @kubawerlos Let me revive this PR. I've rebased this branch, so please review again and let's merge it or suggest what should I do more to be able to merge.

**Comment by Wirone** [comment]
I see a lot of work has been done here and I believe it's worth finishing. @ossinkine is it possible to rebase this again?

@kubawerlos @keradus let's help somehow and finish it 🙂

**Comment by ossinkine** [comment]
@Wirone done, I hope this is the final rebase 😊

**Review by Wirone — APPROVED** [review]
Great job! 🙂

**Comment by Wirone** [comment]
Thank you very much @ossinkine 🍻

**Pull request merged**